### ISSUE TITLE
Exception raised in wrong greenlet

### ISSUE THREAD

**Issue by gambitcostas** [eventlet/eventlet#300]
We have a Django application running inside a Gunicorn eventlet server. When the server starts up, it spawns a greenlet (using eventlet.spawn) which reads data from a socket.

The socket may sometimes get disconnected, in which case we simply log the exception and reconnect again later.

Unfortunately, we noticed the _entire server_ sometimes exits with the following traceback when there's a socket error:
```
Traceback (most recent call last):
  File "/src/env/lib/python3.4/site-packages/gunicorn/arbiter.py", line 515, in spawn_worker
    worker.init_process()
  File "/src/env/lib/python3.4/site-packages/gunicorn/workers/geventlet.py", line 95, in init_process
    super(EventletWorker, self).init_process()
  File "/src/env/lib/python3.4/site-packages/gunicorn/workers/base.py", line 126, in init_process
    self.run()
  File "/src/env/lib/python3.4/site-packages/gunicorn/workers/geventlet.py", line 122, in run
    eventlet.sleep(1.0)
  File "/src/env/lib/python3.4/site-packages/eventlet/greenthread.py", line 34, in sleep
    hub.switch()
  File "/src/env/lib/python3.4/site-packages/eventlet/hubs/hub.py", line 294, in switch
    return self.greenlet.switch()
  File "/src/env/lib/python3.4/site-packages/eventlet/hubs/hub.py", line 332, in run
    self.close_one()
  File "/src/env/lib/python3.4/site-packages/eventlet/hubs/hub.py", line 265, in close_one
    listener.tb(IOClosed(errno.ENOTCONN, "Operation on closed file"))
  File "/src/env/lib/python3.4/site-packages/eventlet/greenthread.py", line 214, in main
    result = function(*args, **kwargs)
  File "/src/apps/home/bookkeep.py", line 180, in run_bookkeeper
    self.data = self.recv_line(30)
  File "/src/apps/home/bookkeep.py", line 164, in recv_line
    raise socket.error("socket is closed")
OSError: socket is closed
[2016-02-14 18:11:27 +0000] [25165] [INFO] Worker exiting (pid: 25165)
```
The code that reads from the socket uses a trampoline to wait for new data, reads, and returns the parsed result. I don't know if the code is relevant (admittedly I didn't write it, and don't particularly like it), but here is the condensed version:
```
def connect(self, timeout=CONNECT_TIMEOUT):
	self._socket = socket.create_connection((self.host, self.port), timeout)

def read_data_from_socket(self, timeout):
	self._socket.setblocking(0)
	deadline = time.time() + timeout
	while not self._received_lines:
		now = time.time()
		try:
			if now >= deadline:
				raise socket.timeout
			try:
				eventlet.hubs.trampoline(self._socket.fileno(),
							read=True, timeout=(deadline - now))
			except eventlet.Timeout:
				continue
			except eventlet.hubs.IOClosed:
				break
			data = self._socket.recv(4096)
			if not data:
				raise socket.error("no data from socket")
		except socket.error as ex:
			raise <some error>
		else:
			if not data:
				eventlet.sleep(0.1)
				continue
			self._received_lines.extend(self.parser.parse(data))

def recv_line(self, timeout=RECEIVE_TIMEOUT):
	# ... code to ensure socket is connected ...
	self.read_data_from_socket(timeout)
	if not self._received_lines:
		raise socket.error("socket is closed")  # this is raised
	return self._received_lines.pop(0)

def bookkeeper(self):  # this is the body of the greenlet spawned on startup
	def run_bookkeeper():
		try:
			self.data = self.recv_line(30)
		except Exception:
			logger.exception("Bookkeeper exception")  # this is logged
			raise
	
	while True:
		eventlet.sleep(75)
		
		task = eventlet.spawn(run_bookkeep)
		try:
			with eventlet.Timeout(30):
				task.wait()
		except:
			logger.error("Bookkeeper failed")  # this is not logged
		else:
			logger.info("Bookkeeper finished")
```
We let Gunicorn monkey-patch for us, and we've verified that the sockets are indeed the green versions. This has happened with both earlier versions of eventlet, and also with 0.18.3.

So, how can the socket error propagate to the "main" Gunicorn loop, in an unrelated greenlet, and cause it to shut down?

As far as we can tell, this appears like a serious bug in either eventlet or greenlet, though even after spending days looking at this we've been unable to make much progress. If anyone has any ideas please let us know!

**Comment by temoto** [comment]
TL;DR: don't do that. When using Eventlet with sockets, you have exactly zero reasons to call either `sock.setblocking(0)` or anything from `eventlet.hubs`, in particular `trampoline()`. 

These lines:
- `self._socket.setblocking(0)`
- `eventlet.hubs.trampoline()`
- `except eventlet.hubs.IOClosed:`

mean author is very well informed about Eventlet internals, because he's recreating parts of it in his application code. So either he had a very specific issue to work around and then we should talk about that issue.

Or, we have a simpler, well tested API that also does what I think your code does:
```
def read_data(self, timeout):
  with eventlet.Timeout(timeout, False):
    while not self.lines:
      data = self.sock.recv(4096)
      if not data:
        return data
      self.lines.extend(self.parser.parse(data))
    return 'ok'
  # log or reconnect or something else
  return 'timeout'

def recv_line(self, timeout):
  if not self.read_data(timeout)
    # socket closed, log or silently reconnect or sys.exit if desire
  return self.lines.pop(0)
```

On a side note, this part is redundant:
```
        task = eventlet.spawn(run_bookkeeper)
            with eventlet.Timeout(30):
                task.wait()
# exactly same behavior:
  with eventlet.Timeout(30):
    run_bookkeeper()
```

**Comment by gambitcostas** [comment]
Thanks temoto for your quick reply. 

I'm positive that that code is much lower-level than it needs to be, and I don't believe it was written this way for any specific reason (I suspect it's very old, and was ported over from code that didn't use greenlets originally). 

What they were aiming for was a continuous processing loop that also polls a socket for new data as it runs but doesn't block if there's nothing to read - so basically EWOULDBLOCK-style behaviour. I agree that what you suggested (plus an extra greenlet) could have achieved the same result and been nicer and more idiomatic.

But anyway, the plan is to remove all of this in the next couple of months as it will be no longer needed, so it's probably not worth the effort to rewrite it right now.. unless you think it might help with the problem we're having?

Tbh, the most pressing issue at the moment is the socket exception _apparently_ being raised in the greenlet that initiated the switch rather than the greenlet being switched to. Are we right in thinking that's what could be happening? Any pointers as to why it might happen or how to debug it would be most appreciated!

**Comment by temoto** [comment]
> not worth the effort to rewrite it right now.. unless you think it might help with the problem we're having?

I'm sorry, didn't make myself clear. I don't think it might help. I'm 100% sure it will help. Playing with internals *is root cause of the problem* you're having.

Yes, you're raising exception into another greenthread. How to debug - very simple, put print with current greenthread id, locals and socket state on each line. Or use ipdb if you prefer interactive debugging.

**Comment by costasgambit** [comment]
Ok, we'll follow your advice. Thanks very much for your help!
